### PR TITLE
[wgsl] Add var initialization tests.

### PR DESCRIPTION
**Pull request by dj2**
This CL adds tests for initializing private and function scoped vars with values.

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [x] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/value_init.spec.ts [hunk 1]
```diff
@@ -0,0 +1,370 @@
+export const description = `Test that variables in the shader are value initialized`;
+
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+import { GPUTest } from '../../gpu_test.js';
+
+export const g = makeTestGroup(GPUTest);
+
+function paramTypeToTestType(ty: string): string {
```
Comment (zoddicus) [inline src/webgpu/shader/execution/value_init.spec.ts:8]: This name is confusing to me, becasue it looks like it is returning a value not a type. Should this be something like `paramValueForTestType` or `valueForTestType`?

**Review by zoddicus — COMMENTED** [review]
Overall looks reasonable, other then one naming thing. Will leave to @dneto0 to do final review, to make sure there are not cases being missed.

**Review by dneto0 — COMMENTED** [review]
Same objection as @zoddicus so we'll let his review govern approval.

Also, the validation tests are to come in another PR, I presume.

File: src/webgpu/shader/execution/value_init.spec.ts (hunk 1 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/value_init.spec.ts:8]: Or just use: `Type[ty].create(1).wgsl()`

File: src/webgpu/shader/execution/value_init.spec.ts [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+  });
+
+g.test('struct')
+  .desc(`Test that structs in private, and function storage classes can be initialized to a value.`)
+  .params(u => u.combine('addressSpace', ['private', 'function'] as const))
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/value_init.spec.ts:321]:
```suggestion
```

File: src/webgpu/shader/execution/value_init.spec.ts [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+    const testScalarValue = paramTypeToTestType(t.params.type);
+
+    let testValue = `${typeDecl}(`;
+    [...Array(arraySize)].map((_, i) => {
+      testValue += `${innerDecl}(`;
+      [...Array(arraySize)].map((_, i) => {
+        testValue += `${testScalarValue},`;
+      });
+      testValue += `),`;
+    });
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/value_init.spec.ts:291]:
```suggestion
    for (let i = 0; i < arraySize; i++) {
      testValue += `${innerDecl}(`;
      for (let j = 0; j < arraySize; j++) {
        testValue += `${testScalarValue},`;
      }
      testValue += `),`;
    }
```

File: src/webgpu/shader/execution/value_init.spec.ts [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+  })
+  .fn(async t => {
+    const arraySize = 4;
+    const typeDecl = `array<${t.params.type}, ${arraySize}>`;
+    const testScalarValue = paramTypeToTestType(t.params.type);
+
+    let testValue = `${typeDecl}(`;
+    [...Array(arraySize)].map((_, i) => {
+      testValue += `${testScalarValue},`;
+    });
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/value_init.spec.ts:246]:
```suggestion
    for (let i = 0; i < arraySize; i++) {
      testValue += `${testScalarValue},`;
    }
```

File: src/webgpu/shader/execution/value_init.spec.ts [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        testScalarValue = '5.0h';
+        break;
+    }
+
+    let testValue = `${typeDecl}(`;
+    [...Array(t.params.c)].map((_, i) => {
+      [...Array(t.params.r)].map((_, k) => {
+        testValue += `${testScalarValue},`;
+      });
+    });
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/value_init.spec.ts:205]:
```suggestion
    for (let c = 0; c < t.params.c; c++) {
      for (let r = 0; r < t.params.r; r++) {
        testValue += `${testScalarValue},`;
      }
    }
```

Comment (dneto0) [reply src/webgpu/shader/execution/value_init.spec.ts:205]: I've seen this go both ways, more C like and more functional-like.

Comment (ben-clayton) [reply src/webgpu/shader/execution/value_init.spec.ts:205]: My objection to this is unnecessary complexity and performance.

* You're constructing an array that needs to be garbage collected.
* For each element of the array, you're calling a function and only using the index.
* That function does some logic, and doesn't return a value.
* The `undefined` value is then used to construct a second array, which is returned and not used. It's another thing we need to GC.

All of this _looks_ like it is doing something more complex than a for loop, when it isn't.

File: src/webgpu/shader/execution/value_init.spec.ts (hunk 1 above)
Comment (dj2) [reply src/webgpu/shader/execution/value_init.spec.ts:8]: What does this do for `bool` map anything non-zero to `true`?

**Comment by dj2** [comment]
K, all comments addressed.

**Review by zoddicus — APPROVED** [review]

**Comment by dj2** [comment]
I checked, the validation tests already exist in `shader,validation,parse,var_and_let`.

**Pull request merged**